### PR TITLE
Implement agent-local revocation mechanism

### PR DESCRIPTION
**Pull request by ueno**
This adds support for the agent-local revocation mechanism as described in the [enhancements#53](https://github.com/keylime/enhancements/blob/master/53_agent_local_revocation.md). It's currently draft state, as it lacks tests and protocol version bump.

### CODE REVIEW

**Comment by kkaarreell** [comment]
Hi @ueno , I think it would be good to have e2e test for this feature. I can create one but just want to confirm. Is changing the settings in `revocation_notifiers` the only thing to be done?

**Comment by ueno** [comment]
> Is changing the settings in revocation_notifiers the only thing to be done?

Yes, I think so. Let me finalize this PR in this week so it can actually be tested.

**Review by ansasaki — CHANGES_REQUESTED** [review]

File: keylime/cloud_verifier_tornado.py
Comment (ueno) [reply keylime/cloud_verifier_tornado.py]: Thanks, fixed.

**Comment by ueno** [comment]
I have no idea why this breaks webhook integration tests, does anyone have any clue?
Also we might need to suppress code scanning warnings on logging certificates.

**Comment by kkaarreell** [comment]
> I have no idea why this breaks webhook integration tests, does anyone have any clue? Also we might need to suppress code scanning warnings on logging certificates.

webhook notifier is not set for some reason. I have added the debug output and only zeromq notifier is in the list.

**Comment by kkaarreell** [comment]
So this is because you have removed `revocation_notifier_webhook` from `keylime.conf` and the function `limeUpdateConf` is only changing existing option, not adding a missing one. I think we need to update the test to set `revocation_notifiers` accordingly.

**Comment by ueno** [comment]
Thanks, but this PR [keeps](https://github.com/keylime/keylime/pull/795/files#diff-bf2d7716986f3aca3950920e063c519a9306b090d95d900bfad8bd5f4c7e6799R32) `revocation_notifier_webhook` for backward compatibility. So if the fallback doesn't take place, I guess that would be an issue in this PR.

**Comment by ueno** [comment]
Ah! That explains it, thanks a lot Karel!

**Comment by ueno** [comment]
@THS-on any suggestion on how to overcome [this](https://github.com/keylime/keylime/pull/795/checks?check_run_id=6687939352) in CodeQL scanning? I don't think revealing certificate path through logging is a problem, but it seems I can't suppress it through `# lgtm` comment (as pre-commit prohibit too long lines...)

**Review by THS-on — APPROVED** [review]
LGTM. The only worry that I have is that an rouge agent can block the async io loop by not responding. Can you test what happens if a agent just never terminates the connection?

**Comment by ueno** [comment]
> The only worry that I have is that an rouge agent can block the async io loop by not responding. Can you test what happens if a agent just never terminates the connection?

That's a good point, thanks for raising it; I'll test it in modified agent with a new e2e test plan. Given that the existing notification mechanism (ZeroMQ) does not guarantee consistency, perhaps we could wrap the `invoke_notify_error` (async) call in a [task](https://docs.python.org/3/library/asyncio-task.html#creating-tasks) to prioritize other awaitables?

**Comment by THS-on** [comment]
> Given that the existing notification mechanism (ZeroMQ) does not guarantee consistency, perhaps we could wrap the invoke_notify_error (async) call in a [task](https://docs.python.org/3/library/asyncio-task.html#creating-tasks) to prioritize other awaitables?

Ideally we would run them in another thread.

**Comment by THS-on** [comment]
> @THS-on, ok I've [wrapped](https://github.com/keylime/keylime/pull/795/commits/0ae060d470b8d9e453b86fca658feee7859e1d9d#diff-67033196f7dec303b3ea55793c6c13fb4b03a72a5515239475a871c3d638dfe4R894-R904) it with threading.Thread (let me know if there is a better approach).

I think it is fine for now.

With zeroMQ all agents are notified when another agent fails. In the current implementation only the failing agent gets message, correct? Is that what we want? I would expect all agents to get the message.

File: keylime/cloud_verifier_tornado.py
Comment (THS-on) [inline keylime/cloud_verifier_tornado.py:871]: response might be `None`.

File: keylime/cloud_verifier_tornado.py
Comment (THS-on) [inline keylime/cloud_verifier_tornado.py:862]: Which part should handle this exception?

File: keylime/cloud_verifier_tornado.py
Comment (THS-on) [inline keylime/cloud_verifier_tornado.py:895]: Why is the loop created here and not in the worker?

**Comment by ueno** [comment]
> With zeroMQ all agents are notified when another agent fails. In the current implementation only the failing agent gets message, correct? Is that what we want? I would expect all agents to get the message.

That's an interesting point I haven't thought about. Perhaps this use-case could be implemented by enabling both "agent" and "webhook" notifiers?

**Comment by THS-on** [comment]
> Perhaps this use-case could be implemented by enabling both "agent" and "webhook" notifiers?

Not really. "webhook" is intended for relaying the information to a single source outside of the Keylime system. Revocation actions are useful for other agents to act on the state change e.g. dropping all connections to that agent. The current implementation just sends this information to the compromised agent which is only useful in some cases.

**Comment by ueno** [comment]
I see; the latest commits should address that (i.e., all the registered agents are notified). I've also changed the usage of threads from manually creating it to use `ThreadPoolExecutor`.

@THS-on could you take a look again?

File: keylime/cloud_verifier_tornado.py
Comment (ueno) [reply keylime/cloud_verifier_tornado.py]: Added a check

File: keylime/cloud_verifier_tornado.py
Comment (ueno) [reply keylime/cloud_verifier_tornado.py]: Yeah, just converted it to a warning.

File: keylime/cloud_verifier_tornado.py
Comment (ueno) [reply keylime/cloud_verifier_tornado.py]: Fixed.

**Comment by THS-on** [comment]
@ueno in general LGTM. Thank you for the changes.

I'll try to test it till the end of the week manually.

**Comment by THS-on** [comment]
During testing I ran into this error message:

```
Jun 18 15:24:49 fedora keylime_verifier[946]: /usr/lib64/python3.10/asyncio/base_events.py:1881: RuntimeWarning: coroutine 'invoke_notify_error' was never awaited
Jun 18 15:24:49 fedora keylime_verifier[946]:   handle = self._ready.popleft()
Jun 18 15:24:49 fedora keylime_verifier[946]: RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

Also can you check on how timeouts are handled if the agent is not available anymore?

**Comment by ueno** [comment]
> During testing I ran into this error message:
> 
> ```
> Jun 18 15:24:49 fedora keylime_verifier[946]: /usr/lib64/python3.10/asyncio/base_events.py:1881: RuntimeWarning: coroutine 'invoke_notify_error' was never awaited
> Jun 18 15:24:49 fedora keylime_verifier[946]:   handle = self._ready.popleft()
> Jun 18 15:24:49 fedora keylime_verifier[946]: RuntimeWarning: Enable tracemalloc to get the object allocation traceback
> ```

Thanks for checking. I was confused with the `loop.run_in_executor()` usage: the result of this function needs to be awaited, as well as `asyncio.as_complete()` has to to be called to wait with timeout (and the results should also be awaited). Those should be fixed now; I have tested with the 4-agent setup described in [keylime-tests](https://github.com/RedHat-SP-Security/keylime-tests/blob/main/TESTING.md#running-multi-host-tests).

> Also can you check on how timeouts are handled if the agent is not available anymore?

Currently, all notification tasks share the same timeout (60 sec).

File: keylime/cloud_verifier_tornado.py
Comment (THS-on) [inline keylime/cloud_verifier_tornado.py:876]: This works only if all the agents event loops are running on a single process. If not you will need to recreate the contexts using `web_util.generate_agent_mtls_context(...)`.

File: keylime/cloud_verifier_tornado.py
Comment (THS-on) [inline keylime/cloud_verifier_tornado.py:71]: This does not work when Keylime is running on multiple processes. See other comment.

**Review by THS-on — CHANGES_REQUESTED** [review]
There is still one issue with the ssl_context, but the rest of the mechanism works nicely.

File: keylime/cloud_verifier_tornado.py
Comment (ueno) [reply keylime/cloud_verifier_tornado.py:864]: Thanks for pointing that out. I've removed the caching but added context recreation for connecting to other agents.

**Review by THS-on — CHANGES_REQUESTED** [review]
One small change then it fully works.

File: keylime/keylime_agent.py
Comment (ueno) [reply keylime/keylime_agent.py:231]: Good catch, fixed

**Review by THS-on — APPROVED** [review]

**Comment by ueno** [comment]
Let me remove the last commit ("Add new test plan that exercises agent-local revocation") which refers to my fork of keylime-tests for testing. Once this PR is merged we can update the upstream keylime-tests repo and then add the test plan back.

**Comment by kkaarreell** [comment]
Hello, is there some work remaining blocking the merge? Since the zeromq removal is important for RHEL-9 packaging and testing, I am trying to find out whether we can expect this change to land within days or later. Thank you.

**Comment by THS-on** [comment]
@kkaarreell there should be nothing blocking this PR. Ideally someone besides me would also review it. @mpeters, @maugustosilva? If it is time sensitive I can merge it, but @mpeters needs to create a new release.

**Comment by kkaarreell** [comment]
It is not time sensitive. I can "workaround it", it just won't be worth if if this PR would be merge in few days.

**Pull request merged**
